---
title: "eQTL Pipeline Report"
output: html_document
params:
  eqtl_results_filtered: NULL
  eqtl_results: NULL
  outdir: NULL
  gds_file: NULL
  single_cell_file: NULL
  counts_assay: NULL
  counts_slot: NULL
  celltype_column: NULL
  individual_column: NULL
  min_cells: NULL
  min_expression: NULL
  cis_distance: NULL
  filter_chr: NULL
  fdr_threshold: NULL
  optimize_pcs: NULL
  optimization_results: NULL
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(data.table)
library(dplyr)
library(ggplot2)
library(tidyr)
format_cis_distance <- function(distance) {
  if (distance >= 1e6) {
    return(paste0(distance / 1e6, "Mb"))
  } else if (distance >= 1e3) {
    return(paste0(distance / 1e3, "Kb"))
  } else {
    return(paste0(distance, "bp"))
  }
}
```

## eQTL Analysis Report

**Report Generated:** `r format(Sys.time(), "%d %B %Y, %H:%M:%S")`

This report was generated using the Nextflow eQTL Pipeline, a comprehensive tool for conducting expression quantitative trait loci (eQTL) analysis on single-cell RNA-seq data.

**GitHub Repository:** [https://github.com/johnsonlab-ic/sceQTL-flow](https://github.com/johnsonlab-ic/sceQTL-flow)

**Contact:** For questions or support regarding this pipeline, please contact Alexander Haglund (GitHub: [HaglundA](https://github.com/HaglundA))

---

### Pipeline Settings

The following settings were used in the pipeline:

- **Output Directory**: `r params$outdir`
- **GDS File**: `r params$gds_file`
- **Single-Cell Data File**: `r params$single_cell_file`
- **Counts Assay**: `r params$counts_assay`
- **Counts Slot**: `r params$counts_slot`
- **Cell-Type Column**: `r params$celltype_column`
- **Individual Column**: `r params$individual_column`
- **Minimum Cells for Pseudobulking**: `r params$min_cells`
- **Minimum Expression Percentage for Genes**: `r params$min_expression`
- **Cis Distance for eQTL Analysis**: `r format_cis_distance(params$cis_distance)`
- **Chromosomes tested**: `r params$filter_chr`
- **FDR Threshold for eQTL Results**: `r params$fdr_threshold`
- **Optimize Principal Components**: `r params$optimize_pcs`

```{r echo=FALSE}
"""
# Load the filtered eQTL results
eqtl_results_filtered <- readRDS(params$eqtl_results_filtered)

# Load the eQTL results
eqtl_results <- readRDS(params$eqtl_results)

# Number and names of cell-types tested
cell_types <- names(eqtl_results_filtered)
num_celltypes <- length(cell_types)

# Average number of genes tested
genes_list <- lapply(eqtl_results, function(df) unique(df$gene))
all_genes <- unlist(genes_list)
avg_genes_tested <- length(all_genes) / num_celltypes

# Number of SNPs tested
snps_list <- lapply(eqtl_results, function(df) unique(df$SNP))
all_snps <- unique(unlist(snps_list))
num_snps_tested <- length(all_snps)
"""

---

## PC Optimization Results

```{r child='../quarto_reports/pc_optimization_report.qmd'}
```
```

### Input Overview

- **Number of cell-types tested:** `r num_celltypes`
- **Names of cell-types tested:** `r paste(cell_types, collapse=", ")`
- **Average number of genes tested per cell-type:** `r avg_genes_tested`
- **Number of SNPs tested:** `r num_snps_tested`

`r if (params$optimize_pcs) "### PC optimization results"`
```{r echo=FALSE}
#load pc optimization results
if (params$optimize_pcs) {
  # Load pc optimization results
  file_list <- params$optimization_results
  file_list <- unlist(strsplit(file_list, " "))
  results <- rbindlist(lapply(file_list, fread))
  results$celltype <- factor(results$celltype)

  # Plot pc optimization results
  plot <- ggplot(results, aes(x = n_pcs, y = n_egenes)) +
      geom_point(aes(fill = celltype), size=0.8, shape=21) +
      geom_line(aes(color=celltype)) +
      facet_wrap(~ celltype) +
      xlab("Number of Principal Components") +
      ylab("Number of eGenes") +
      theme_minimal()+theme(legend.position = "none")

  print(plot)
}
```


### Total and Unique eGenes (5% FDR)

```{r plot_total_unique_eGenes, echo=FALSE}
total_genes <- function(cell_list) {
  sapply(cell_list, function(df) length(unique(df$gene)))
}

total_unique_genes <- function(cell_list) {
  sapply(cell_list, function(df) {
    # Genes in the current data frame
    current_genes <- unique(df$gene)
    
    # Genes in other data frames
    other_genes <- unique(unlist(sapply(cell_list, function(other_df) {
      if (identical(df, other_df)) return(NULL)
      return(other_df$gene)
    })))
    
    # Genes that are unique to the current data frame
    length(setdiff(current_genes, other_genes))
  })
}

mateqtlouts <- readRDS(params$eqtl_results_filtered)

mateqtlouts_filt <- lapply(mateqtlouts, function(x) {
  x <- x %>% 
    mutate(se = beta / t.stat, fstat = t.stat^2) 
  top_snps <- x %>%
    group_by(gene) %>%
    arrange(p.value) %>%
    slice_head(n = 1) %>%
    ungroup() %>%
    arrange(FDR)
  return(top_snps)
})

# Get total eGenes (aggregate)
tot_eGenes <- total_genes(mateqtlouts_filt)
tot_unique_eGenes <- total_unique_genes(mateqtlouts_filt)
df <- data.frame(tot_eGenes, tot_unique_eGenes)
df$celltype <- rownames(df)
df$celltype <- gsub("_", " ", df$celltype)

# Get unique eGenes 
color_pal <- scales::hue_pal()(length(unique(df$celltype)))
colorvec <- setNames(color_pal, unique(df$celltype))

df$colorvec <- colorvec
tot_eGenes_sum <- lapply(mateqtlouts_filt, function(x) {
  return(x$gene)
})
tot_eGenes_sum <- length(unique(unlist(tot_eGenes_sum)))

df <- df[order(df$tot_eGenes, decreasing = TRUE), ]
df_long <- df %>%
  pivot_longer(cols = c(tot_eGenes, tot_unique_eGenes), names_to = "gene_type", values_to = "count")
df_long$gene_type <- factor(df_long$gene_type, levels = c("tot_unique_eGenes", "tot_eGenes"))
df_long$celltype_col <- paste0(df_long$celltype, df_long$gene_type)
df_long$celltype_col <- rep(c(2, 1), times = nrow(df_long) / 2)
df_long$celltype_col <- factor(df_long$celltype_col)
df_long$celltype <- factor(df_long$celltype, levels = rev(df$celltype))
df_long$celltype_col <- paste0(df_long$celltype, df_long$celltype_col)

colorvec <- df_long$colorvec
names(colorvec) <- df_long$celltype_col

make_lighter <- function(hex, fraction = 0.4) {
  rgb <- col2rgb(hex)
  new_rgb <- rgb + fraction * (255 - rgb)
  return(rgb(new_rgb[1], new_rgb[2], new_rgb[3], maxColorValue = 255))
}

colorvec <- df_long$colorvec
new_colors <- c(rbind(colorvec[c(TRUE, FALSE)], sapply(colorvec[c(TRUE, FALSE)], make_lighter)))
names(new_colors) <- df_long$celltype_col
max_val <- max(df_long$count) + 0.12 * max(df_long$count) 

g <- ggplot(df_long, aes(x = celltype, y = count, fill = celltype_col)) +
  geom_bar(stat = 'identity', position = position_dodge(), width = 0.8, colour = "black") +
  geom_text(aes(x = celltype, y = count, label = count), position = position_dodge(width = 0.9), hjust = -0.2, size = 6 / .pt) +
  scale_fill_manual(values = new_colors) +
  labs(y = "Total vs unique eGenes (5% FDR)") +
  coord_flip(ylim = c(NA, max_val)) +
  theme_classic() +
  theme(axis.text.x = element_text(size = 6.5, colour = "black"),
        axis.title.y = element_blank(),
        axis.text.y = element_text(size = 6.5, colour = "black"),
        axis.title.x = element_text(size = 7, face = "bold"),
        legend.key.size = unit(0.7, "cm"),
        legend.position = "none",
        legend.text = element_text(size = 7))

print(g)
```
